File: app/src/stores/modulations.ts
import type {
  Modulation,
  ModulationSerialized,
  Modulator,
  Optional,
} from '@/types'
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { useDevice } from './device'
import { useBridge } from '@/bridge'
import { map as mapValue, unpackBytes } from '@/utils'
import { useEventBus } from '@vueuse/core'
import { useItems } from './items'

type Id = Modulation['id']

export const useModulations = defineStore('modulations', () => {
  const map = ref(new Map<Id, Modulation>())

  const device = useDevice()
  const bridge = useBridge()
  const items = useItems()

  const modulatorValueBus = useEventBus('modulator-value')
  bridge.on('/n/modulations/update', ({ args }) => {
    for (let i = 0; i < args.length; i += 2) {
      const info = args[i]
      const value = args[i + 1]

      const type = info & 0xff
      if (type === 0) {
        const [, modulatorId] = unpackBytes(info)
        modulatorValueBus.emit(modulatorId, value)
      } else if (type === 1) {
        const [, itemId, propIndex] = unpackBytes(info)
        const item = items.instances.get(itemId)
        if (!item) continue

        const definition = items.definitions.get(item.type)
        if (!definition) continue

        const propName = Object.keys(definition.props).find(
          (name) => definition.props[name].options.index === propIndex,
        )
        if (!propName) continue

        item.modulatedProps[propName] = value
      }
    }
  })

  // Getters
  const getByProp = (itemId: number, prop: string) =>
    computed(() =>
      Array.from(map.value.values()).find(
        (item) => item.itemId === itemId && item.prop === prop,
      ),
    )

  const list = computed(() => Array.from(map.value.values()))

  const getByItemId = (id: number) =>
    computed(() => list.value.filter(({ itemId }) => itemId === id))

  // Todo: make this computed as well
  const getByModulatorId = (id: Modulator['id']) =>
    list.value.filter((item) => item.modulatorId === id)

  // Actions
  const serialize = (): ModulationSerialized[] =>
    Array.from(map.value.values()).map(
      ({ modulatorId, itemId, prop, amount }) => [
        modulatorId,
        itemId,
        prop,
        amount,
      ],
    )

  const deserialize = (serialized: ModulationSerialized[]) => {
    map.value.clear()
    serialized.forEach(([modulatorId, itemId, prop, amount]) =>
      add({ modulatorId, itemId, prop, amount }),
    )
  }

  const add = (modulation: Optional<Modulation, 'id'>, updateDevice = true) => {
    modulation.id ??= `${modulation.itemId}-${modulation.prop}`
    map.value.set(modulation.id, modulation as Modulation)

    if (updateDevice) {
      device.update('/r/modulations/add', [
        modulation.modulatorId,
        modulation.itemId,
        modulation.prop,
        modulation.amount,
      ])
    }
  }

  const remove = (id: Id, updateDevice = true) => {
    const modulation = map.value.get(id)
    if (!modulation) return

    map.value.delete(id)

    if (updateDevice) {
      const { modulatorId, itemId, prop } = modulation
      device.update('/r/modulations/remove', [modulatorId, itemId, prop])
    }
  }

  const updateAmount = (
    id: Id,
    amount: Modulation['amount'],
    updateDevice = true,
  ) => {
    const modulation = map.value.get(id)
    if (!modulation) return

    modulation.amount = amount
    if (updateDevice) {
      const { modulatorId, itemId, prop } = modulation
      device.update('/r/modulations/amount', [
        modulatorId,
        itemId,
        prop,
        amount,
      ])
    }
  }

  const clear = () => map.value.clear()

  return {
    map,
    getByProp,
    getByItemId,
    getByModulatorId,
    serialize,
    deserialize,
    add,
    remove,
    updateAmount,
    clear,
  }
})

File: engine/src/Modulations.lua
Modulations = {
  ---@type Modulation[]
  list = {},
}

---@param serialized ModulationSerialized[]
function Modulations.deserialize(serialized)
  for _, modulation in ipairs(serialized) do
    -- https://github.com/LuaLS/lua-language-server/issues/1353
    ---@diagnostic disable-next-line: param-type-mismatch
    Modulations.add(unpack(modulation))
  end
end

---@param item Item
---@param prop string
function Modulations.getByItem(item, prop)
  for _, modulation in pairs(Modulations.list) do
    if modulation[2] == item and modulation[3] == prop then
      return modulation
    end
  end
end

---@param modulatorId number
---@param itemId number
---@param prop string
---@param amount number
function Modulations.add(modulatorId, itemId, prop, amount)
  local modulator = Items.instances[modulatorId] --[[@as Modulator]]
  if not modulator then error(Log.messageItemNotFound(modulatorId)) end

  local item = Items.instances[itemId]
  if not item then error(Log.messageItemNotFound(itemId)) end

  Modulations.list[#Modulations.list + 1] = { modulator, item, prop, amount }
  Miwos:emit('patch:change')
end

---@param modulatorId number
---@param itemId number
---@param prop string
function Modulations.remove(modulatorId, itemId, prop)
  for i, modulation in pairs(Modulations.list) do
    if
      modulation[1].__id == modulatorId
      and modulation[2].__id == itemId
      and modulation[3] == prop
    then
      Modulations.list[i] = nil
      Miwos:emit('patch:change')
      return
    end
  end
end

function Modulations.clear()
  Modulations.list = {}
  Miwos:emit('patch:change')
end

---@param modulatorId number
---@param itemId number
---@param prop string
---@param amount number
function Modulations.updateAmount(modulatorId, itemId, prop, amount)
  for _, modulation in pairs(Modulations.list) do
    if
      modulation[1].__id == modulatorId
      and modulation[2].__id == itemId
      and modulation[3] == prop
    then
      modulation[4] = amount
      return
    end
  end
end

function Modulations.update(time, updateApp)
  local modulatorValues = {}
  local updateSerialized = {}

  for id, item in pairs(Items.instances) do
    if item.__definition.category == 'modulators' then
      ---@cast item Modulator

      local value = item:value(time)
      modulatorValues[id] = value

      updateSerialized[#updateSerialized + 1] = Utils.packBytes(0, id)
      updateSerialized[#updateSerialized + 1] = value
    end
  end

  for _, modulation in pairs(Modulations.list) do
    local modulator, item, prop, amount = unpack(modulation)
    -- https://github.com/LuaLS/lua-language-server/issues/135
    ---@cast modulator -string, -number, -Item
    ---@cast item Item

    if modulator and item then
      local type, options = unpack(item.__definition.props[prop])
      local definition = Miwos.definitions.props[type]

      local baseValue = item.props.__values[prop]
      local modulationValue = modulatorValues[modulator.__id]
      local value =
        definition.modulateValue(baseValue, modulationValue, amount, options)

      Items.updateModulatedProp(item, prop, value)
      updateSerialized[#updateSerialized + 1] =
        Utils.packBytes(1, item.__id, options.index)
      updateSerialized[#updateSerialized + 1] = value
    end
  end

  if Utils.option(updateApp, true) and #updateSerialized > 0 then
    Bridge.notify('/n/modulations/update', unpack(updateSerialized))
  end
end
